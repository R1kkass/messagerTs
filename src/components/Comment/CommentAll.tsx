import { domen } from "../../Const/Const";
import { FC, useEffect, useState } from "react";
import { IUnitComment } from "./Comment";
import { fetchComment } from "./CommentService";
import { Link } from "react-router-dom";

interface IAll{
    comment: IUnitComment[]
}

const CommentAll:FC<IAll> = ({comment})=>{
    const [widthCom, setWidthCom] = useState<boolean>(false)
    const [comments, setComments] = useState<IUnitComment[]>([])
    const allComment =async ()=>{
        
        const resp = await fetchComment(comment[0]?.feedId)
        setComments(resp)
        setWidthCom(true)
    }

    useEffect(()=>{
        if(comments.length){
            allComment()
        }
    }, [comment])
    
    if(widthCom){
        return (
            <div>
                {comments?.map((comment)=>(
                    <div key={comment.id} className="Comment__unit">
                        <div className="Unit__Img">
                            <img src={`http://${domen}/${comment.user.img}`} alt="" />
                        </div>
                        <div>
                            <div className="Unit__name">
                                <Link to={`/my/${comment.user.id}`}>{comment.user.name}</Link>
                            </div>
                            <div dangerouslySetInnerHTML={{__html: comment.text}} className="Unit__text" />
                        </div>
                        
                    </div>
                ))}
            </div>
        )
    }


    return (
        <div>
            
            {comment?.map((comment)=>(
                <div key={comment.id} className="Comment__unit">
                    <div className="Unit__Img">
                        <img src={`http://${domen}/${comment.user.img}`} alt="" />
                    </div>
                    <div>
                        <div className="Unit__name">
                            <Link to={`/my/${comment.user.id}`}>{comment.user.name}</Link>
                        </div>
                        <div dangerouslySetInnerHTML={{__html: comment.text}} className="Unit__text">
                        </div>
                    </div>
                    
                </div>
            ))}
            {comment.length >= 2 ?  <p className="Comment_ShowAll" onClick={allComment}>Показать всё</p> : ''}
           
        </div>
    )
}

export default CommentAll