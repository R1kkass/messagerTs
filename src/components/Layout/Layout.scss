@import "../../Styles/Variables.scss";

.Layout{
    height: 60px;
    
    &__fixed{
        z-index: 9;
        position: fixed;
        height: 60px;
        width: 100%;
        color: white;
        background: $main-color;
        display: flex;
        align-items: center;
        padding-left: 20px;
        padding-right: 20px;
        display: flex;
        justify-content: space-between;
    }
    .Link{
        background-color: rgb(24, 24, 134);
        color: white;
        padding: 10px;
        border-radius: 5px;
        text-decoration: none;
        transition: 0.3s ;
        margin-left: 10px;
        
        border: none;
        outline: none;
        &:hover{
            border-radius: 10px;
            background-color: white;
            color: $main-color;
        }
        
    }
    .activeLink{
        background-color: white;
        color: $main-color;
        border-radius: 5px;
        text-decoration: none;
        padding: 10px;
        transition: 0.3s;
        margin-left: 10px;
        &:hover{
            border-radius: 10px;
        }
    }
}