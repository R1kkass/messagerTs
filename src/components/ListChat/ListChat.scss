@import '../../Styles/Variables.scss';

.ListChat{
    &__block{
        overflow: hidden;
        height: 60px;
        
        display: flex;
        font-size: 16px;

        justify-content: space-around;
        padding-left: 20px;
        align-items: center;
        border: solid 1px $main-color;
        background-color:  rgb(105, 156, 252);
        cursor: pointer;
        vertical-align: center;
        a{
            text-decoration: none;
            color: black;
            color: white;
            display: grid;
            grid-template-columns: 70px 1fr;
            width: 100%;
            align-items: center;
            height: 100%;
        }
        &:hover{
            background-color: rgb(136, 174, 243);

        }
    }
    &__img{
        display: flex;
        
    }
    .Img__lastUser{
        width: 30px;
        height: 30px;
        border-radius: 50%;
        margin-right: 10px;
        object-fit: cover;
    }
    .Img__Creator{
        object-fit: cover;
        width: 50px;
        height: 50px;
        border-radius: 50%;
        margin-right: 10px;
    }
}