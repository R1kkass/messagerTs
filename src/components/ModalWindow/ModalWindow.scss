@import '../../Styles/Variables.scss';

@keyframes slideIn {
    0%{
        opacity: 0;
    }
    100%{
        opacity: 1;
    }
}

.ModaWindow__button{
    margin: 0 auto;
}

%ModalWindow{
    position: fixed;
    bottom: 0;
    left: 0;
    right: 0;
    top: 0;
    min-width: 100%;
    min-height: 100vh;
    background-color: none;
    display: flex;
    justify-content: center;
    align-items: center;
    .ModalWindow__children{
        width: 50%;
        height: 600px;
        min-width: 300px;
        overflow: hidden;
        background-color: $main-color;
        border-radius: 20px;
        z-index: 999;
        opacity: 1;
    }
    .ModalWindow__window{
        z-index: 1;
        width: 100%;
        height: 100vh;
        position: fixed;
        background: rgba(136, 136, 136, 0.2);
    }
}

.ModalWindow{
    @extend %ModalWindow;
}


.my-node-enter {
    
    @extend %ModalWindow;
    opacity: 0;
  }
  .my-node-enter-active {

    opacity: 1;
    transition: opacity 500ms;
  }
  .my-node-exit {
    
    opacity: 1;
  }
  .my-node-exit-active {
    opacity: 0;
    transition: opacity 500ms;
  }