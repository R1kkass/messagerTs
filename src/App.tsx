import Websoket from './Page/Websocket/Websoket';
import React, { useEffect } from 'react'
import { BrowserRouter, Route, Routes } from 'react-router-dom';
import Counter from './Page/Counter';
import { useDispatch, useSelector } from 'react-redux';
import * as firebase from 'firebase/app'
import * as firestore from 'firebase/firestore'
import * as auth from 'firebase/auth'
import Login from './Page/Login/Login';
import Layout from './components/Layout/Layout';
import { initializeApp} from 'firebase/app';
import MyRouter from './Page/Routes/MyRoutes';

function App() {
  


  return (
      <div className="App">
        <BrowserRouter>
          <Layout />
          <MyRouter />
          </BrowserRouter>
      </div>
  );
}

export default App;
