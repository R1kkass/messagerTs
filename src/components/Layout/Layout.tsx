import React,{FC, memo, useCallback, useEffect, useMemo, useReducer, useState} from "react"
import { Button } from "react-bootstrap"
import {useAuthState} from 'react-firebase-hooks/auth'
import { useDispatch, useSelector } from "react-redux";
import {getAuth, GoogleAuthProvider, signInWithPopup, signOut} from 'firebase/auth'
import './Layout.scss'
import { Link, NavLink, useLocation, useParams } from "react-router-dom";
import jwtDecode from "jwt-decode";
import { check, checkAuth } from "./LayoutService";
import { tokenAction } from "../../store/token";

export interface IUserReducer {
    auth: {
        auth: any,
        firebase: any,
        firestore: any
    }
}

const Layout:FC = memo(()=>{
    const toks = useSelector((state:any)=>state.token.token)
    console.log(toks);
    const [tok, setTok] = useState(toks)
    const token: any= toks 
    const location = useLocation()
    const dispatch =useDispatch()
    

    useEffect(()=>{
        check()
        .finally(()=>{
            setTok(localStorage.getItem('token'))
        })
        dispatch(tokenAction())
    },[location, tok])



    if(!tok){
    return(
    <div className="Layout">
        
        <div className="Layout__fixed">
        <NavLink  
            className={({ isActive }) =>
            isActive ? 'activeLink' : "Link"}
            to="/login">
            Вход
        </NavLink>
        <NavLink  
            className={({ isActive }) =>
            isActive ? 'activeLink' : "Link"}
            to="/registration">
            Регистрация
        </NavLink>
        </div>
    </div> 
    )}

    return(
        <div className="Layout">
            <div className="Layout__fixed">
                <NavLink
                className={({ isActive }) =>
                isActive ? 'activeLink' : "Link"}
                to={`/my/${toks?.id}`}>
                    {toks?.email}
                </NavLink>
               <button className="Link" onClick={()=>{localStorage.removeItem('token');setTok(null); check()}}>Выйти</button>
            </div>
        </div>
    )
})

export default Layout