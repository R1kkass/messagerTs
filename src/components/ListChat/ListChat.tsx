import axios from "axios";
import { domen, URi } from "../../Const/Const";
import jwtDecode from "jwt-decode";
import React, { memo, useEffect, useRef } from "react";
import { useDispatch, useSelector } from "react-redux";
import { Link } from "react-router-dom";
import { chatAction } from "../../store/chat";
import { IReduceState, IUnitChat } from "../../types/IReduce";
import './ListChat.scss'

interface IListChat{
    data:{
        user: IList[]
    }
}

interface IList{
    secondUser: string,
    userCreator: string,
    user:string,
    lastUser: string,
    idRoom:string,
    lastMessage:string
}

const ListChat = memo(()=>{
    
    const dispatch = useDispatch()
    const socket = useRef<WebSocket | null>(null)
    const listChat:IListChat = useSelector((state:IReduceState)=>state.chat.chats)

   

 const user:any =localStorage.getItem('token') ? jwtDecode(localStorage.getItem('token') || '') : localStorage.getItem('token')    
 async function fetchChat(){
        let tokensMain: any = jwtDecode(localStorage.getItem('token') || '')
        const response = await axios.get(`${URi}/chat/getall?email=${tokensMain.email}`)
        await dispatch(chatAction(response))
        
        return response
    }


    useEffect(()=>{
        fetchChat()
    },[])
    
    function connect(){
        socket.current = new WebSocket("ws://localhost:5001/con")
        
        socket.current.onopen = ()=>{
          const message = {
              event: "connectionChat",
              username: user?.email
            }
            socket.current?.send(JSON.stringify(message))
            setInterval(()=>{
    
            },15000)
        }
        
        socket.current.onmessage = (e)=>{    
            fetchChat()
        }
        socket.current.onerror = ()=>{
          console.log('ошибка');
          
      }
      socket.current.close = ()=>{
        console.log('Подключение отключено');
    }
      }
    
    useEffect(()=>{
      connect()
    },[])

    if(!localStorage.getItem('token')){
        return(
            <div>Не авторизован</div>
        )
    }

    return(
        <div  className="ListChat">
            {listChat?.data?.user?.map((list:any)=>(
                <div data-testid="ListChat" className="ListChat__block">
                    <Link to = {`/im/${list?.idRoom}`}>
                    {list?.secondUser!==user?.email ? 
                    <>
                        <img className="Img__Creator" 
                            src={`http://${domen}/${list?.secondUser}.jpg`}
                            alt={list?.user}/>
                            <div>
                            <div className="ListChat__secondUser">{list?.secondUser}</div>
                            <div className="ListChat__img">
                                <img
                                className="Img__lastUser"
                                src={`http://${domen}/${list?.lastUser}.jpg`}
                                alt="" />
                                <div>{list?.lastMessage}</div>
                            </div>
                        </div>
                    </>
                    :
                    <>
                    <img className="Img__Creator" 
                    src={`http://${domen}/${list?.userCreator}.jpg`}
                    alt={list?.user}/>
                    <div>
                        <div className="ListChat__secondUser">{list?.userCreator}</div>
                        <div className="ListChat__img" >
                            <img
                            className="Img__lastUser" 
                            src={`http://${domen}/${list?.lastUser}.jpg`} 
                            alt=""
                            />
                            <div>{list?.lastMessage}</div>
                        </div>
                    </div>
                </>
                    }
                    
                    </Link>
                </div>
            ))}
        </div>
    )
})

export default ListChat