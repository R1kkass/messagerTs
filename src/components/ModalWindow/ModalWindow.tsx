import React, { FC, useRef, useState } from "react";
import { Button } from "react-bootstrap";
import { CSSTransition } from "react-transition-group";
import "./ModalWindow.scss"

interface IModalWindow{
    children?: React.ReactNode
}

const ModalWindow:FC<IModalWindow> = ({children})=>{

    const [visible,setVisible] = useState(false)

    const nodeRef = useRef<any>()

    
    return(
    <>
      <Button className="ModalWindow__button" onClick={()=>setVisible(true)}>Изменить аватар</Button>
            <CSSTransition nodeRef={nodeRef} in={visible} timeout={500} classNames="my-node" unmountOnExit
        onEnter={() => setVisible(true)}
        onExited={() => setVisible(false)}>            
            <div className={'ModalWindow'} ref={nodeRef}>
                <div className="ModalWindow__window" onClick={()=>setVisible(false)}>
                    
                </div>
                    <div className="ModalWindow__children">
                            {children}
                    </div>
            </div>
        </CSSTransition>
        </>
    )
}

export default ModalWindow