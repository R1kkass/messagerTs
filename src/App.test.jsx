import React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';
import News from './Page/News/News';
import NewsUpdate from 'components/NewsBlock/NewsUpdate';
import App from 'App';
import Layout from 'components/Layout/Layout';
import VisibleWindow from 'components/VisibleWindow/VisibleWindow';
import { Provider } from 'react-redux';
import { store } from 'store';
import ListChat from 'components/ListChat/ListChat';
import axios from 'axios';
import { MemoryRouter, Route, Router, Routes } from 'react-router-dom';
import { resp, responseCommentTest, subTestArr, visibleTestArr } from 'testArray/testArray';
import LaftBlockAllSub from 'components/LeftBlockMain/LeftBlockAllSub';
import ModalWindow from 'components/ModalWindow/ModalWindow';
import CommentAll from 'components/Comment/CommentAll';
import { fetchNewsUnit } from 'Page/News/NewsService';
import userEvent from '@testing-library/user-event';

jest.mock('axios')

describe('TEST Components', ()=>{

  let response:any;
  beforeEach(()=>{
    response=resp
  })

  test('ListChat TEST',async () => {
    localStorage.setItem('token','eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpZCI6MSwiZW1haWwiOiJkb2dpbkB5YW5kZXgucnUiLCJyb2xlIjoiVVNFUiIsImltZyI6ImRvZ2luQHlhbmRleC5ydS5qcGciLCJuYW1lIjoi0JDQvdC00YDQtdC5INCU0L7Qs9C40L0iLCJpYXQiOjE2NzM2MjQ5OTksImV4cCI6MTY3MzcxMTM5OX0.itvGfnTXfvCnBlNyzNla1LfVwU4tEzS5wYPrPZsPp8I')
    await axios.get.mockReturnValue(response)
    render(<Provider store = {store}><ListChat/></Provider>, {wrapper: MemoryRouter});
    const list = await screen.findAllByTestId('ListChat');
    expect(list.length).toBe(2)
    })

  let responses:any;
  beforeEach(()=>{
    responses = visibleTestArr
  })

  test('VisibleWindow TEST',async () => {
    axios.get.mockReturnValue(responses)
    render(<Provider store = {store}><VisibleWindow visible={true} /></Provider>);
    const user = await screen.findAllByTestId('visble');
    expect(user.length).toBe(3)
    })

    let responseSubs:any = subTestArr
  
    test('LeftSub TEST',async () => {
      axios.get.mockReturnValue(responseSubs)
      render(<Provider store = {store}><LaftBlockAllSub /></Provider>);
      const btn = screen.getByTestId('toggle-btn')
      fireEvent.click(btn)
      const user = await screen.findAllByTestId('subTest');
      
      expect(user.length).toBe(3)
      })  
      
      test('Modal TEST',async () => {
        axios.get.mockReturnValue(responseSubs)
        render(<Provider store = {store}><ModalWindow nameButton='test' /></Provider>);
        const btn = screen.getByTestId('toggle-btn')
        fireEvent.click(btn)
        const modal = screen.getByTestId('modalTest')
        expect(modal).toBeInTheDocument()
        })


      test('CommentAll TEST',async () => {
        axios.get.mockReturnValue(responseCommentTest || '')
        render(<Provider store = {store}><CommentAll comment={responseCommentTest} /></Provider>);
        const btn = screen.getByTestId('toggle-btn')
        fireEvent.click(btn)
        const user = await screen.findAllByTestId('subTest');
        
        expect(user.length).toBe(3)
      })
});