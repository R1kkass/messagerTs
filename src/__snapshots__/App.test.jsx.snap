// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TEST Components Modal TEST 1`] = `
<div
  class="ModalWindow my-node-exit my-node-exit-active"
  data-testid="modalTest"
>
  <div
    class="ModalWindow__window"
    data-testid="modalHidden"
  />
  <div
    class="ModalWindow__children"
  />
</div>
`;
